### PR TITLE
Lint system catalog

### LINKED ISSUE #19563: Add a lint to validate that the documentation of introspection sources matches their schema
### Feature request

At the moment, the [documentation](https://github.com/MaterializeInc/materialize/blob/ef389ee7ed42be96b177d68a72a1e0c4e50bc547/doc/user/content/sql/system-catalog/mz_internal.md) of introspection sources is separate from their [definition](https://github.com/MaterializeInc/materialize/blob/ef389ee7ed42be96b177d68a72a1e0c4e50bc547/src/adapter/src/catalog/builtin.rs). This causes them to diverge frequently. We should add a lint that compares the schema to the documentation and errors if it doesn't match.

Alternatively, we could consider an approach where both definition and documentation are generated from a common source.

### PR DESCRIPTION
**Pull request by antiguru**
Lint the system catalog documentation against the implementation. This helps to ensure that the documentation and implementation stay in sync.

### Motivation

Fixes #19563.

Documentation changes are best viewed with whitespace changes hidden!

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: doc/user/content/sql/system-catalog/mz_internal.md [hunk 1]
```diff
@@ -326,234 +345,308 @@ Per-worker relations expose the same data as their global counterparts, but have
...
+| `export_id`  | [`text`]    | The ID of the dataflow export. Corresponds to [`mz_compute_exports.export_id`](#mz_compute_exports).                                                                                                                                                 |
+| `import_id`  | [`text`]    | The ID of the dataflow import. Corresponds to either [`mz_catalog.mz_sources.id`](../mz_catalog#mz_sources) or [`mz_catalog.mz_tables.id`](../mz_catalog#mz_tables) or [`mz_catalog.mz_materialized_views.id`](../mz_catalog#mz_materialized_views). |
+| `delay_ns`   | [`uint8`]   | The upper bound of the bucket in nanoseconds.                                                                                                                                                                                                        |
+| `count`      | [`numeric`] | The (noncumulative) count of delay measurements in this bucket.                                                                                                                                                                                      |
```
Comment (nrainer-materialize) [inline doc/user/content/sql/system-catalog/mz_internal.md:393]: It is at least surprising to see `count` having type `numeric`... 🤔 
(Though, I understand this change just aligns documentation with reality.)

A bit below, `sent` and `received` (messages) are a similar case.

File: ci/test/lint-docs-catalog.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  doc/user/content/sql/system-catalog/mz_catalog.md
+)
+
+slt_directory=test/sqllogictest/autogenerated
+
+rewrite=false
+
+if [[ "${1:-}" = --rewrite ]]; then
+    shift
+    rewrite=true
```
Comment (nrainer-materialize) [inline ci/test/lint-docs-catalog.sh:34]: 👍 for the rewrite functionality

File: ci/test/pipeline.template.yml [hunk 3]
```diff
@@ -124,6 +124,15 @@ steps:
       queue: linux-x86_64
     coverage: skip
 
+  - id: lint-docs-catalog
+    label: Lint catalog docs
+    command: bin/ci-builder run stable ci/test/lint-docs-catalog.sh
+    inputs: [doc/user]
```
Comment (nrainer-materialize) [inline ci/test/pipeline.template.yml:130]: Docs deviate when either the docs or the code change. I understand that we don't want to run this on every PR, but what about running it additionally in nightly?

**Review by def- — APPROVED** [review]
Python code seems fine, lint runs through quickly.

File: doc/user/content/sql/system-catalog/mz_catalog.md [hunk 4]
```diff
@@ -401,19 +433,21 @@ Field                  | Type                         | Meaning
 
 The `mz_system_privileges` table contains information on system privileges.
 
+<!-- RELATION_SPEC mz_catalog.mz_system_privileges -->
 Field         | Type     | Meaning
 --------------|----------|--------
-`privileges` | [`mz_aclitem array`] | The privileges belonging to the system.
```
Comment (antiguru) [inline doc/user/content/sql/system-catalog/mz_catalog.md:406]: @jkosh44 FYI -- I'm not sure what is true!

File: doc/user/content/sql/system-catalog/mz_catalog.md [hunk 5]
```diff
@@ -457,3 +494,5 @@ Field          | Type                 | Meaning
 [`uint8`]: /sql/types/uint8
 [`uint4`]: /sql/types/uint4
 [`mz_aclitem array`]: /sql/types/mz_aclitem
+
+<!-- RELATION_SPEC_UNDOCUMENTED mz_catalog.mz_operators -->
```
Comment (antiguru) [inline doc/user/content/sql/system-catalog/mz_catalog.md:498]: @morsapaes We're not documenting `mz_operators`, should we?

File: doc/user/content/sql/system-catalog/mz_internal.md [hunk 6]
```diff
@@ -300,14 +319,14 @@ messages and additional metadata helpful for debugging.
...
-| ------------------------ | ------------------------------ | --------                                                                                                                   |
-| `id`                     | [`text`]                       | The ID of the subscription.                                                                                                |
-| `session_id`             | [`uint4`]                      | The ID of the session that runs the subscription. Corresponds to [`mz_sessions.id`](#mz_sessions).                         |
-| `cluster_id`             | [`text`]                       | The ID of the cluster on which the subscription is running. Corresponds to [`mz_clusters.id`](../mz_catalog/#mz_clusters). |
-| `created_at`             | [`timestamp with time zone`]   | The time at which the subscription was created.                                                                            |
-| `referenced_object_ids`  | [`text list`]                  | The IDs of objects referenced by the subscription. Corresponds to [`mz_objects.id`](../mz_catalog/#mz_objects)             |
```
Comment (antiguru) [inline doc/user/content/sql/system-catalog/mz_internal.md:309]: @antiguru `text list` is correct.

File: doc/user/content/sql/system-catalog/mz_internal.md (hunk 1 above)
Comment (antiguru) [reply doc/user/content/sql/system-catalog/mz_internal.md:393]: @antiguru to file a bug.

File: doc/user/content/sql/system-catalog/mz_internal.md [hunk 7]
```diff
@@ -326,234 +345,308 @@ Per-worker relations expose the same data as their global counterparts, but have
...
-| --------------- | ------------ | -------                                                  |
-| `slept_for_ns`  | [`bigint`]   | The actual length of the park event in nanoseconds.      |
-| `requested_ns`  | [`bigint`]   | The requested length of the park event in nanoseconds.   |
-| `count`         | [`bigint`]   | The (noncumulative) count of park events in this bucket. |
+<!-- RELATION_SPEC mz_internal.mz_scheduling_parks_histogram -->
+| Field           | Type        | Meaning                                                  |
+| --------------- |-------------| -------                                                  |
+| `slept_for_ns`  | [`uint8`]   | The actual length of the park event in nanoseconds.      |
+| `requested_ns`  | [`uint8`]   | The requested length of the park event in nanoseconds.   |
+| `count`         | [`numeric`] | The (noncumulative) count of park events in this bucket. |
```
Comment (antiguru) [inline doc/user/content/sql/system-catalog/mz_internal.md:646]: @antiguru unexpected `numeric`

**Review by tr0njavolta — APPROVED** [review]

File: doc/user/content/sql/system-catalog/mz_internal.md (hunk 1 above)
Comment (antiguru) [reply doc/user/content/sql/system-catalog/mz_internal.md:393]: #20522

File: ci/test/pipeline.template.yml [hunk 8]
```diff
@@ -124,6 +124,15 @@ steps:
...
+    timeout_in_minutes: 30
```
[the remark this answers is not in the document]
Comment (antiguru) [reply ci/test/pipeline.template.yml]: Integrated with `lint-docs.sh`.

File: doc/user/content/sql/system-catalog/mz_catalog.md [hunk 9]
```diff
@@ -422,24 +456,27 @@ Field        | Type                 | Meaning
...
 
+<!-- RELATION_SPEC mz_catalog.mz_types -->
 Field          | Type                 | Meaning
 ---------------|----------------------|----------
 `id`           | [`text`]             | Materialize's unique ID for the type.
 `oid`          | [`oid`]              | A [PostgreSQL-compatible OID][oid] for the type.
-`schema_id`    | [`uint8`]            | The ID of the schema to which the type belongs. Corresponds to [`mz_schemas.id`](/sql/system-catalog/mz_catalog/#mz_schemas).
+`schema_id`    | [`text`]             | The ID of the schema to which the type belongs. Corresponds to [`mz_schemas.id`](/sql/system-catalog/mz_catalog/#mz_schemas).
 `name`         | [`text`]             | The name of the type.
+`category`     | [`text`]             | The category of the type.
```
[the remark this answers is not in the document]
Comment (antiguru) [reply doc/user/content/sql/system-catalog/mz_catalog.md:466]: #20523

File: doc/user/content/sql/system-catalog/mz_catalog.md (hunk 5 above)
Comment (antiguru) [reply doc/user/content/sql/system-catalog/mz_catalog.md:498]: #20524

File: doc/user/content/sql/system-catalog/mz_internal.md [hunk 10]
```diff
@@ -569,3 +662,17 @@ The `mz_scheduling_parks_histogram` view describes a histogram of [dataflow] wor
 [`timestamp with time zone`]: /sql/types/timestamp
 [arrangement]: /get-started/arrangements/#arrangements
 [dataflow]: /get-started/arrangements/#dataflows
+
+<!-- RELATION_SPEC_UNDOCUMENTED mz_internal.mz_aggregates -->
```
Comment (antiguru) [inline doc/user/content/sql/system-catalog/mz_internal.md:666]: #20526

File: doc/user/content/sql/system-catalog/mz_internal.md [hunk 11]
```diff
@@ -326,234 +345,308 @@ Per-worker relations expose the same data as their global counterparts, but have
...
-| `channel_id`       | [`bigint`]   | The ID of the channel. Corresponds to [`mz_dataflow_channels.id`](#mz_dataflow_channels). |
-| `sent`             | [`bigint`]   | The number of messages sent.                                                              |
-| `received`         | [`bigint`]   | The number of messages received.                                                          |
+<!-- RELATION_SPEC mz_internal.mz_message_counts -->
+| Field              | Type        | Meaning                                                                                   |
+| ------------------ |-------------| --------                                                                                  |
+| `channel_id`       | [`uint8`]   | The ID of the channel. Corresponds to [`mz_dataflow_channels.id`](#mz_dataflow_channels). |
+| `sent`             | [`numeric`] | The number of messages sent.                                                              |
+| `received`         | [`numeric`] | The number of messages received.                                                          |
```
Comment (nrainer-materialize) [inline doc/user/content/sql/system-catalog/mz_internal.md:578]: I was wondering about the type of these as well and guess they also should not be `numeric`.

Comment (antiguru) [reply doc/user/content/sql/system-catalog/mz_internal.md:578]: Yup, I agree! Listed in #20522.

**Pull request merged**